### PR TITLE
Fix `copy_transpose_valid` condition for `copy_same_type_transpose_`

### LINKED ISSUE #64358: copy_ produces wrong results for large enough dst + broadcasting
## 🐛 Bug

Mathematically, for a matrix `A`, `A ** 0` is the identity matrix.
For tensors `A` of shape `(a, b, 1, 1)` for large enough `a, b`, the result of `torch.linalg.matrix_power(A, n=0)` is not `torch.ones(a, b, 1, 1)`, but contains arbitrary values.

## To Reproduce

```python
import torch

a = 32
b = 256
A = torch.randn(torch.randn(a, b, 1, 1))
A_p = torch.linalg.matrix_power(A, 0)

assert torch.all(A_p == torch.ones(a, b, 1, 1))
```

## Expected behavior

Assertion goes through: `A_p` contains only 1s.

## Environment

```
Collecting environment information...
PyTorch version: 1.9.0
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.9.6 | packaged by conda-forge | (default, Jul 11 2021, 03:39:48)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-81-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 2080
Nvidia driver version: 460.91.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.2
[pip3] torch==1.9.0
[pip3] torchvision==0.2.2
[conda] blas                      2.111                       mkl    conda-forge
[conda] blas-devel                3.9.0            11_linux64_mkl    conda-forge
[conda] cudatoolkit               11.1.74              h6bb024c_0    nvidia
[conda] libblas                   3.9.0            11_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            11_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            11_linux64_mkl    conda-forge
[conda] liblapacke                3.9.0            11_linux64_mkl    conda-forge
[conda] mkl                       2021.3.0           h726a3e6_557    conda-forge
[conda] mkl-devel                 2021.3.0           ha770c72_558    conda-forge
[conda] mkl-include               2021.3.0           h726a3e6_557    conda-forge
[conda] numpy                     1.21.2           py39hdbf815f_0    conda-forge
[conda] pytorch                   1.9.0           py3.9_cuda11.1_cudnn8.0.5_0    pytorch
[conda] torchvision               0.2.2                      py_3    pytorch
```

## Additional context

I only found the issue for the trailing dimensions (1, 1).
`np.linalg.matrix_power(A, n=0)` behaves correctly.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @mruberry @rgommers @heitorschueroff

### PR DESCRIPTION
**Pull request by IvanYashchuk**
Thanks to @ngimel for the hint where the problem might be (https://github.com/pytorch/pytorch/issues/64358#issuecomment-910868849)!

I added a test that fails on master to verify the fix. The shape `(60, 60)` was chosen because of `MIN_SZ = 60 * 60` in `copy_transpose_valid`.

Fixes https://github.com/pytorch/pytorch/issues/64358

### CODE REVIEW

**Comment by ngimel** [comment]
That's a correct fix, but I've also ran a few benchmarks, and it turns out that at least until 1000x1000 (I didn't check further) special transpose implementation is worse than regular copy (I've compared against (1,x,x) copies that do absolutely same work with same strides, but won't trigger special path because dim check will fail, `t=Timer(stmt="a.view(1,size,size).copy_(b.view(1,size,size))", globals=globals())`), so probably we should remove this special path altogether?

**Comment by IvanYashchuk** [comment]
This is interesting. In my tests, the special transpose copy is much better than the regular one for sizes larger than 128x128.

```
                                |    copy    |  copy_view  |  transpose_copy
 -----------------------------------------------------------------
      torch.Size([59, 59])      |       2.5  |        4.3  |         2.4    
      torch.Size([60, 60])      |       2.2  |        4.0  |         3.0    
      torch.Size([128, 128])    |      15.8  |       17.9  |        11.1    
      torch.Size([256, 256])    |      60.2  |       64.3  |        38.6    
      torch.Size([512, 512])    |     578.4  |      572.8  |       146.4    
      torch.Size([1024, 1024])  |    5641.0  |     5669.4  |       582.3    
      torch.Size([2048, 2048])  |   21209.0  |    21369.0  |      4449.3    
      torch.Size([4096, 4096])  |  123850.9  |   123991.5  |     21192.4    
      torch.Size([8192, 8192])  |  484488.8  |   486212.8  |     94623.4    

Times are in microseconds (us).

```

<details>
  <summary>Script used for benchmarking</summary>
  
I've tested with `OMP_NUM_THREADS=1` and without it, it doesn't affect the numbers. Ryzen Threadripper 3970X was used.
  
```python
import torch
import time
from torch.utils.benchmark import Timer
from torch.utils.benchmark import Compare
import sys

print('Using pytorch %s' % (torch.__version__))

shapes = [59, 60, 128, 256, 512, 1024, 2048, 4096, 8192]
results = []
num_threads = 1
dtype = torch.float32
repeats = 2

for shape in shapes:
    a = torch.randn(shape, shape)
    b = torch.randn(shape, shape)
    b.transpose_(-2, -1)
    assert b.transpose(-2, -1).is_contiguous()

    av = a.clone().reshape(1, shape, shape)
    bv = b.clone().reshape(1, shape, shape)
    assert bv.transpose(-2, -1).is_contiguous()

    tasks = [("av.copy_(bv)", "copy"), ("a.view(1,shape,shape).copy_(b.view(1,shape,shape))", "copy_view"), ("a.copy_(b)", "transpose_copy"),]
    timers = [Timer(stmt=stmt, num_threads=num_threads, label=f"copy", sub_label=f"{a.shape}", description=label, globals=globals()) for stmt, label in tasks]

    for i, timer in enumerate(timers * repeats):
        results.append(
            timer.blocked_autorange()
        )
        print(f"\r{i + 1} / {len(timers) * repeats}", end="")
        sys.stdout.flush()

compare = Compare(results)
# compare.trim_significant_figures()
compare.colorize()
compare.print()

```
  
</details>

**Comment by ngimel** [comment]
Weird, I see a big jump for 
```
a=torch.randn(size, size)
b=a.clone() 
t=Timer(stmt="a.view(1,size,size).copy_(b.t().view(1,size,size))", globals=globals())
```
when size goes from 1000 to 1024. For 1000 it's 900 us, for 1024 it's 3.5 ms. But anyway, that's outside of scope for this PR, special transpose kernel is still beneficial, so let's land as is.

**Review by ngimel — APPROVED** [review]
